pkg/provisioner/ironic/ironic.go

```diff
@@ -1295,3 +1338,53 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// softPowerOff sends 'soft power off' request to BM node.
+// If soft power off is not supported, the request ends with an error.
+// Otherwise the request ends with no error and the result should be
+// checked later via node fields "power_state" and "last_error".
+func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error) {
+	p.log.Info("ensuring host is powered off by \"soft power off\" command")
+
+	ironicNode, err := p.findExistingHost()
+	if err != nil {
+		return result, errors.Wrap(err, "failed to find existing host")
+	}
+
+	if ironicNode.PowerState != powerOff {
+		// Although request is "soft power off", target state is still "power off"
+		if ironicNode.TargetPowerState == powerOff {
+			p.log.Info("waiting for power status to change")
+			result.RequeueAfter = powerRequeueDelay
+			result.Dirty = true
+			return result, nil
+		}
+		// Checks last error of ironic node, if there is one happened
+		// then soft power off has failed
+		if p.host.Status.SoftPowerOffStatus == metal3v1alpha1.StateSoftPowerOffTriggered {
```

Comment:
If the Ironic API docs are correct, then we don't need to store this status in the CRD. Avoiding that would be preferable if it's possible.

Reply:
So the docs were incorrect.
Still, with no local state required, couldn't we do something like:
* get the status
* if the target power state is not `power off`:
  * start a soft power off with timeout
* else (target state is `power off`):
  * if there's an error:
    * start a hard power off
  * else (no error):
    * keep waiting

How could that fail? The only way I can think of is the periodic power state check in Ironic overwriting the error before we see it. But if our timeout << power state check period, then I think we should always end up doing the Right Thing eventually.

Reply:
I change to check the power_state and target_power_state instead of last_error. It seems if soft power off fails, target_power_state will be unset.